lib/Basic/RedirectedProcess.cpp

```diff
@@ -0,0 +1,217 @@
+  sigemptyset(&mostSignals);
+  for (int i = 1; i < SIGUNUSED; ++i) {
+    if (i == SIGKILL || i == SIGSTOP)
+      continue;
+    sigaddset(&mostSignals, i);
+  }
+  posix_spawnattr_setsigdefault(&attributes, &mostSignals);
+#else
+  sigset_t mostSignals;
+  sigfillset(&mostSignals);
+  sigdelset(&mostSignals, SIGKILL);
+  sigdelset(&mostSignals, SIGSTOP);
+  posix_spawnattr_setsigdefault(&attributes, &mostSignals);
+#endif
+
+  // Establish a separate process group.
+  posix_spawnattr_setpgroup(&attributes, 0);
+
+  // Set the attribute flags.
+  unsigned flags = POSIX_SPAWN_SETSIGMASK | POSIX_SPAWN_SETSIGDEF;
+  if (setGroupFlags)
+    flags |= POSIX_SPAWN_SETPGROUP;
+
+// Close all other files by default.
+//
+// FIXME: This is an Apple-specific extension, and we will have to do
+// something else on other platforms (and unfortunately, there isn't
+// really an easy answer other than using a stub executable).
+#ifdef __APPLE__
+  flags |= POSIX_SPAWN_CLOEXEC_DEFAULT;
+#endif
+
+  posix_spawnattr_setflags(&attributes, flags);
+
+  // Setup the file actions.
+  posix_spawn_file_actions_t fileActions;
+  posix_spawn_file_actions_init(&fileActions);
+
+  // Open /dev/null as stdin.
+  posix_spawn_file_actions_addopen(&fileActions, 0, "/dev/null", O_RDONLY, 0);
+
+  // Create a pipe to use to read the command output, if necessary.
+  if (shouldCaptureOutput) {
+    // Open the write end of the pipe as stdout and stderr.
+    posix_spawn_file_actions_adddup2(&fileActions, innerProcessInfo->writeHandle, 1);
+    posix_spawn_file_actions_adddup2(&fileActions, innerProcessInfo->writeHandle, 2);
+
+    // Close the read and write ends of the pipe.
+    posix_spawn_file_actions_addclose(&fileActions, innerProcessInfo->readHandle);
+    posix_spawn_file_actions_addclose(&fileActions, innerProcessInfo->writeHandle);
+  } else {
+    posix_spawn_file_actions_adddup2(&fileActions, 1, 1);
+    posix_spawn_file_actions_adddup2(&fileActions, 2, 2);
+  }
+
+  // Spawn the command.
+  pid_t pid;
+  {
+    // We need to hold the spawn processes lock when we spawn, to ensure that
+    // we don't create a process in between when we are cancelled.
+    std::lock_guard<std::mutex> guard(spawnedProcessesMutex);
+
+    if (posix_spawn(&pid, path, /*file_actions=*/&fileActions,
+                    /*attrp=*/&attributes, args, envp) != 0) {
+      return false;
+    }
+  }
+
+  posix_spawn_file_actions_destroy(&fileActions);
+  posix_spawnattr_destroy(&attributes);
+
+  innerProcessInfo->processId = pid;
+
+  return true;
+}
+
+bool RedirectedProcess::readPipe(llvm::SmallString<1024> &output) {
+  if (!shouldCaptureOutput) {
+    return true;
+  }
+
+  // Close the write end of the output pipe.
+  ::close(innerProcessInfo->writeHandle);
+
+  // Read all the data from the output pipe.
+  while (true) {
+    char buf[4096];
+    ssize_t numBytes = ::read(innerProcessInfo->readHandle, buf, sizeof(buf));
+    if (numBytes < 0) {
+      return false;
```

Comment:
We should loop here if `errno == EINTR` (again another problem in existing code)